bitrisescript/src/bitrisescript/bitrise.py

```diff
@@ -0,0 +1,218 @@
+    """Waits for the given Bitrise build_slug to finish.
+
+    Args:
+        build_slug (str): The Bitrise build to wait on.
+        poll_interval (int): Time in seconds to poll the build (default: 60).
+    """
+    client = BitriseClient()
+    endpoint = f"/builds/{build_slug}"
+
+    while True:
+        response = await client.do_request(endpoint)
+        data = response["data"]
+
+        if data["finished_at"]:
+            log.info(f"Build {build_slug} is finished")
+            break
+
+        log.debug(f"Build {build_slug} is still running, waiting {poll_interval}s...")
+        await asyncio.sleep(poll_interval)
+
+    if data["status_text"] != "success":
+        if data["status_text"] == "error":
+            raise Exception(f"Job {build_slug} errored! Got: {response}")
+        if data["status_text"] == "aborted":
+            raise Exception(f"Job {build_slug} was aborted. Got: {response}")
+        raise Exception(f"Job {build_slug} is finished but not successful. Got: {response}")
+
+
+async def download_artifacts(build_slug: str, artifacts_dir: str) -> None:
+    """Download all artifacts from the given Bitrise build.
+
+    Args:
+        build_slug (str): The Bitrise build to download artifacts from.
+        artifacts_dir (str): Directory to save artifacts to.
+    """
+    client = BitriseClient()
+    endpoint = f"/builds/{build_slug}/artifacts"
+    response = await client.do_request(endpoint)
+
+    artifacts_metadata = {metadata["slug"]: metadata["title"] for metadata in response["data"]}
+
+    for artifact_slug, title in artifacts_metadata.items():
+        endpoint = f"/builds/{build_slug}/artifacts/{artifact_slug}"
+        data = (await client.do_request(endpoint))["data"]
+
+        download_url = data["expiring_download_url"]
+        await download_file(download_url, os.path.join(artifacts_dir, title))
+
+
+async def download_log(build_slug: str, artifacts_dir: str, poll_interval: int = 60) -> None:
+    """Download the log from the given Bitrise build.
+
+    Args:
+        build_slug (str): The Bitrise build to download the log from.
+        artifacts_dir (str): Directory to save artifacts to.
+        poll_interval (int): Time in seconds to poll for completion of log (default: 60).
+    """
+    client = BitriseClient()
+    endpoint = f"/builds/{build_slug}/log"
+
+    while True:
+        response = await client.do_request(endpoint)
+        if response["is_archived"] is True:
+            log.info("Log is now ready")
+            break
+        else:
+            log.info("Log is still running. Waiting another minute...")
+            await asyncio.sleep(poll_interval)
+
+    download_url = response["expiring_raw_log_url"]
+    if download_url:
+        await download_file(download_url, os.path.join(artifacts_dir, "bitrise.log"))
+    else:
+        log.error("Bitrise has no log to offer for job {0}. Please check https://app.bitrise.io/build/{0}".format(build_slug))
```

Comment:
To be honest, I'm not sure. This logic was copy/pasted from existing logic in `firefox-ios`. I'm not sure if the log being missing means there was some fatal error or not. Maybe there's cases where the log doesn't exist but other artifacts do? But that seems unlikely.

I think you're right and it would be safer to just raise an exception here, and if we ever run into a scenario where it's expected to not have a log, we can adjust this at that time.

Reply:
Actually, on second look the log url expires in 10 minutes, so maybe that's the motivation behind making it an error rather than an exception. Would we really want to hold up a whole graph just because we missed our chance to download the bitrise log? The error message even links to the build where you can presumably still see the bitrise log, so now I'm thinking this shouldn't be fatal.